pkg/providers/snow/reconciler/reconciler.go

```diff
@@ -2,19 +2,151 @@ package reconciler
+		log.Error(err, "Failed to set up snow env vars")
+		return controller.Result{}, err
+	}
+
+	if !dataCenterConfig.Status.SpecValid {
+		log.Info("Skipping cluster reconciliation because datacenter config is invalid", "data center", dataCenterConfig.Name)
+		return controller.Result{
+			Result: &ctrl.Result{
+				Requeue:      true,
+				RequeueAfter: defaultRequeueTime,
+			},
+		}, nil
+	}
+
+	log.V(4).Info("Fetching cluster spec")
+	kubeClient := clientutil.NewKubeClient(r.client)
+	specWithBundles, err := c.BuildSpec(ctx, kubeClient, cluster)
+	if err != nil {
+		return controller.Result{}, err
+	}
+
+	configManager := snow.NewConfigManager(r.defaulters, r.validator)
+	if err = configManager.SetDefaultsAndValidate(ctx, specWithBundles.Config); err != nil {
+		return controller.Result{}, err
+	}
+	log.Info("cluster", "name", cluster.Name)
+
+	if result, err := r.reconcileControlPlaneSpec(ctx, cluster, specWithBundles, kubeClient, log); err != nil {
+		return result, err
+	}
+
+	capiCluster, result, errCAPICLuster := r.getCAPICluster(ctx, cluster, log)
+	if errCAPICLuster != nil {
+		return result, errCAPICLuster
+	}
+
+	if !conditions.IsTrue(capiCluster, controlPlaneReadyCondition) {
+		log.Info("waiting for control plane to be ready", "cluster", capiCluster.Name, "kind", capiCluster.Kind)
+		return controller.Result{Result: &ctrl.Result{
+			RequeueAfter: defaultRequeueTime,
+		}}, err
+	}
+
 	return controller.Result{}, nil
 }
+
+func (r *Reconciler) reconcileControlPlaneSpec(ctx context.Context, cluster *anywherev1.Cluster, specWithBundles *c.Spec, kubeClient *clientutil.KubeClient, log logr.Logger) (controller.Result, error) {
+	if !conditions.IsTrue(cluster, controlSpecPlaneAppliedCondition) {
+		log.Info("Applying control plane spec", "name", cluster.Name)
+		controlPlaneSpec, err := snow.ControlPlaneSpec(ctx, specWithBundles, kubeClient)
+		if err != nil {
+			return controller.Result{}, err
+		}
+		if err := serverside.ReconcileYaml(ctx, r.client, controlPlaneSpec); err != nil {
+			return controller.Result{Result: &ctrl.Result{
+				RequeueAfter: defaultRequeueTime,
+			}}, err
+		}
+		conditions.MarkTrue(cluster, controlSpecPlaneAppliedCondition)
+	}
+	return controller.Result{}, nil
+}
+
+func (r *Reconciler) getCAPICluster(ctx context.Context, cluster *anywherev1.Cluster, log logr.Logger) (*clusterv1.Cluster, controller.Result, error) {
+	capiCluster := &clusterv1.Cluster{}
+	capiClusterName := types.NamespacedName{Namespace: constants.EksaSystemNamespace, Name: cluster.Name}
+	log.Info("Searching for CAPI cluster", "name", cluster.Name)
+	if err := r.client.Get(ctx, capiClusterName, capiCluster); err != nil {
+		return nil, controller.Result{Result: &ctrl.Result{
+			Requeue:      true,
+			RequeueAfter: defaultRequeueTime,
+		}}, err
+	}
+	return capiCluster, controller.Result{}, nil
+}
+
+func setupEnvVars(ctx context.Context, cli client.Client) error {
+	credentials, caBundle, err := getSnowCredentials(ctx, cli)
```

Comment:
```suggestion
	if credentials, caBundle, err := getSnowCredentials(ctx, cli); err != nil {
```

To keep it consistent with the code below